src/meta/src/hummock/hummock_manager.rs

```diff
@@ -133,6 +133,7 @@ macro_rules! commit_multi_var {
 
 #[derive(Default)]
 struct Versioning {
+    checkpoint_version_id: u64,
     current_version_id: CurrentHummockVersionId,
     hummock_versions: BTreeMap<HummockVersionId, HummockVersion>,
     hummock_version_deltas: BTreeMap<HummockVersionId, HummockVersionDelta>,
```

Comment:
Currently we keep both the `HummockVersionDelta` and the full `HummockVersion` in memory for all pinned version id, which looks redundant to me. I think we only need to keep three things in memory:
1. The persisted full `HummockVersion`
2. The current `HummockVersion` (optional, it can be re-constructed by 1) and 3))
3. `HummockVersionDelta` between 1) and 2).

```
| HummockVersion           |  -> persisted, in-memory
| VersionDelta1            |  -> persisted, in-memory
| VersionDelta2            |  -> persisted, in-memory
      ...
| VersionDeltaX            |  -> persisted, in-memory
| HummockVersion (current) |  -> in-memory

```

Reply:
Btw, `stale_sstables` is also a duplicate of `hummock_version_deltas`.

Reply:
> Currently we keep both the `HummockVersionDelta` and the full `HummockVersion` in memory for all pinned version id, which looks redundant to me. I think we only need to keep three things in memory:
> 
> 1. The persisted full `HummockVersion`
> 2. The current `HummockVersion` (optional, it can be re-constructed by 1) and 3))
> 3. `HummockVersionDelta` between 1) and 2).
> 
> ```
> | HummockVersion           |  -> persisted, in-memory
> | VersionDelta1            |  -> persisted, in-memory
> | VersionDelta2            |  -> persisted, in-memory
>       ...
> | VersionDeltaX            |  -> persisted, in-memory
> | HummockVersion (current) |  -> in-memory
> ```

will be fixed in following PRs

Reply:
> 



> Currently we keep both the `HummockVersionDelta` and the full `HummockVersion` in memory for all pinned version id, which looks redundant to me. I think we only need to keep three things in memory:
> 
> 1. The persisted full `HummockVersion`
> 2. The current `HummockVersion` (optional, it can be re-constructed by 1) and 3))
> 3. `HummockVersionDelta` between 1) and 2).
> 
> ```
> | HummockVersion           |  -> persisted, in-memory
> | VersionDelta1            |  -> persisted, in-memory
> | VersionDelta2            |  -> persisted, in-memory
>       ...
> | VersionDeltaX            |  -> persisted, in-memory
> | HummockVersion (current) |  -> in-memory
> ```

but we will need to reconstruct version from lots of deltas every time `pin_version` is called

Reply:
After we implement a mechanism for CN to only fetch version deltas to pin a version, it is rare that we need to reconstruct versions in meta. We can revisit this after we have the incremental fetch implemented.

Reply:
> Btw, `stale_sstables` is also a duplicate of `hummock_version_deltas`.

`stale_sstables` is not real `stale_sstables`. Currently it saves tables which are staled and not marked as deleted. `stale_sstables` will be updated every time a SST is marked as deleted. Therefore, we can neither reconstruct versions from `stale_sstables` nor replace `stable_sstables` with version deltas.

Reply:
The reason why we need `stale_sstables` previously is because we don't have an efficient way to lookup deleted ssts for a given version, which is exactly what version delta stores, so `stale_sstables` can be replaced by version delta.

> stale_sstables will be updated every time a SST is marked as deleted

I think here you mean we remove a `stale_sstables` entry when the corresponding SST is marked for deletion. This step is not a requirement since we can clear `stale_sstables` when the version is GCed. You can refer to LevelDB/RocksDB on how   version delta (or VersionEdit in LevelDB/RocksDB's term) is being used and how SST deletion works.